src/QUICConnection.ts

```diff
@@ -515,507 +545,753 @@ class QUICConnection extends EventTarget {
+
+      const localError = this.conn.localError();
+      if (localError == null) {
+        // This is a software error
+
+        // const e_ = new errors.ErrorQUICConnectionInternal(
+        //   'Failed `recv` with unknown internal error',
+        //   { cause: e }
+        // );
+        // this.conn.close(
+        //   false,
+        //   ConnectionErrorCode.InternalError,
+        //   Buffer.from('')
+        // );
+        // this.dispatchEvent(
+        //   new events.EventQUICConnectionError({
+        //     detail: e_
+        //   })
+        // );
+        // this.dispatchEvent(
+        //   new events.EventQUICConnectionClose({
+        //     detail: {
+        //       type: 'local',
+        //       ...this.conn.localError()!
+        //     }
+        //   })
+        // );
+
+        // The problem is that both "caller error" and "software" error
+        // Is executed by throwing it upwards
+        // It is up the caller to distinguish between the 2
+        // That if it is a software error, it just has to keep bubbling up
+        // If it is a caller error, it may be able to handle it
+        // If it is caller error, it may bea domain error on the caller
+        throw e;
+      } else {
+        // This is a legitimate state transition of the connection
+        // So it is not a caller error, therefore we do not throw it up
+        const e_ = new errors.ErrorQUICConnectionLocal(
+          'Failed connection due to local error',
+          {
+            cause: e,
+            data: localError,
+          }
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionError({ detail: e_ })
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionClose({
+            detail: {
+              type: 'local',
+              ...localError
+            }
+          })
+        );
         return;
       }
-
-      if (this.conn.isInEarlyData() || this.conn.isEstablished()) {
-        await this.processStreams();
-      }
-    } finally {
-      this.logger.debug('RECV FINALLY');
-      if (
-        this[status] !== 'destroying' &&
-        (this.conn.isClosed() || this.conn.isDraining())
-      ) {
-        this.logger.debug('calling stop due to closed or draining');
-        // Destroy in the background, we still need to process packets.
-        // Draining means no more packets are sent, so streams must be force closed.
-        void this.stop({ force: true }, mon).catch(() => {});
-      }
     }
+    // If `config.verifyCallback` is not defined, simply being established is
+    // sufficient to mean we are securely established, however if it is defined
+    // then secure establishment occurs only after custom TLS verification has
+    // passed.
+    if (this.conn.isEstablished() && this.config.verifyCallback != null) {
+      this.resolveSecureEstablishedP();
+    }
+    // If we are "secure established" we can process streams.
```

Comment:
Logic error, if the  `this.config.verifyCallback` is *not* defined then we resolve.

Reply:
Yes, so the current fix ensures that if `verifyCallback` is set, then it will run the check under `send`, and if it isn't set, it will resolve the `secureEstablishedP` if the connection passed establishment.